play-validations/memory-footprint/src/main/java/com/google/wear/watchface/dfx/memory/ArscTable.java

```diff
@@ -0,0 +1,129 @@
+ *
+ * Note that more than one ArscResource object can exist for the given dimensions. For example, if
+ * there is a drawable and a drawable-fr folder, then there may be multiple ArscResource entries for
+ * drawables with the same type, name and extension. The configuration detail, e.g. "fr" or
+ * "default", is not currently exposed in the ArscResource objects as it isn't used.
+ */
+public class ArscTable {
+    // Only certain resource types are of interest to the evaluator, notably, not string resources.
+    private static final Set<String> RESOURCE_TYPES = Set.of("raw", "xml", "drawable", "font", "asset");
+    private static final String RESOURCES_FILE_NAME = "resources.arsc";
+    private final List<ArscResource> resourceTable;
+
+    private ArscTable(List<ArscResource> resources) {
+        this.resourceTable = resources;
+    }
+
+    /**
+     * Creates the table from a path to an AAB structure on the file system.
+     * @param aabPath The path to the root of the AAB directory.
+     * @return The constructed table.
+     * @throws IOException when the resources file cannot be found, or other IO errors occur.
+     */
+    static ArscTable createFromAabDirectory(Path aabPath) throws IOException {
+        Path arscPath = Paths.get(aabPath.toString(), RESOURCES_FILE_NAME);
+        File arscFile = new File(arscPath.toAbsolutePath().toString());
+        if (!arscFile.exists()) {
+            throw new FileNotFoundException("Resources file not found");
+        }
+        InputStream is = new FileInputStream(arscFile);
+        return createTable(is, null, aabPath);
+    }
+
+    /**
+     * Creates the table from a AAB or APK file.
+     * @param zipFile The zip file object representing the AAB/APK.
+     * @return The constructed table.
+     * @throws IOException when the resources file cannot be found, or other IO errors occur.
+     */
+    static ArscTable createFromAndroidPackage(ZipFile zipFile) throws IOException {
+        ZipEntry arscEntry = new ZipEntry(RESOURCES_FILE_NAME);
+        InputStream is = zipFile.getInputStream(arscEntry);
+        return createTable(is, zipFile, null);
+    }
+
+    /**
+     * Creates the table from an InputStream.
+     *
+     * @param is The InputStream
+     * @param zipFile The source zip file. This must be populated when creating the table from a AAB
+     *                or APK archive. Where the filesystem approach is taken, this should be null.
+     * @param aabRoot The source AAB directory path. This should be null when creating the table
+     *                from an AAB/APK archive.
+     * @return The constructed table.
+     * @throws IOException when errors loading resources occur.
+     */
+    private static ArscTable createTable(InputStream is, ZipFile zipFile, Path aabRoot) throws IOException {
+        BinaryResourceFile resources = BinaryResourceFile.fromInputStream(is);
+        ResourceTableChunk table = (ResourceTableChunk) resources.getChunks().get(0);
```

Comment:
Should errors here be handled gracefully?

https://cs.android.com/android-studio/platform/tools/base/+/mirror-goog-studio-main:apkparser/cli/src/main/java/com/android/tools/apk/analyzer/ApkAnalyzerImpl.java;l=193

Reply:
Nice! Thanks! Done.